File: featurify-grpc-client/src/main/java/ru/a1pha1337/featurify/client/FeaturifyClient.java
package ru.a1pha1337.featurify.client;

import ru.a1pha1337.featurify.grpc.proto.BooleanFeatureResponse;
import ru.a1pha1337.featurify.grpc.proto.EnumFeatureResponse;
import ru.a1pha1337.featurify.grpc.proto.PayloadFeatureResponse;

/** Blocking, thread-safe client. RPC failures retain their gRPC status and trailers. */
public interface FeaturifyClient {
    /** Returns a JSON document. The default preserves compatibility with existing custom clients. */
    default PayloadFeatureResponse getPayloadFeature(String key, String group) {
        throw io.grpc.Status.UNIMPLEMENTED.withDescription("PAYLOAD is not supported by this client").asRuntimeException();
    }

    default PayloadFeatureResponse getPayloadFeature(String key) {
        return getPayloadFeature(key, null);
    }

    BooleanFeatureResponse getBooleanFeature(String key, String group);

    default BooleanFeatureResponse getBooleanFeature(String key) {
        return getBooleanFeature(key, null);
    }

    EnumFeatureResponse getEnumFeature(String key, String group);

    default EnumFeatureResponse getEnumFeature(String key) {
        return getEnumFeature(key, null);
    }

    BooleanFeatureResponse getVectorFeature(String key, String element, String group);

    default BooleanFeatureResponse getVectorFeature(String key, String element) {
        return getVectorFeature(key, element, null);
    }
}

File: featurify-grpc-starter/src/main/kotlin/ru/a1pha1337/featurify/client/FeaturifyClient.kt
package ru.a1pha1337.featurify.client

import ru.a1pha1337.featurify.grpc.proto.BooleanFeatureResponse
import ru.a1pha1337.featurify.grpc.proto.EnumFeatureResponse

/** Blocking, thread-safe client. RPC failures propagate as gRPC StatusRuntimeException, including trailers. */
interface FeaturifyClient {
    fun getBooleanFeature(
        key: String,
        group: String? = null,
    ): BooleanFeatureResponse

    fun getEnumFeature(
        key: String,
        group: String? = null,
    ): EnumFeatureResponse

    fun getVectorFeature(
        key: String,
        element: String,
        group: String? = null,
    ): BooleanFeatureResponse
}
